### PR TITLE
Cg0562

### PR DESCRIPTION
**Pull request by SFJohnson24**
test data:
[pos.json](https://github.com/user-attachments/files/23463339/pos.json)
[Rule_underscores.json](https://github.com/user-attachments/files/23510741/Rule_underscores.json)
[neg.json](https://github.com/user-attachments/files/23463338/neg.json)

This PR adds wildcard logic for all operations as well as allowing the regex argument = for the above rule.  

This pull request introduces improvements to how domain wildcards (like `--SEQ`) are replaced with concrete domain-specific variable names (like `AESEQ`) across the codebase, especially in rule processing and operation parameter handling. It also adds support for the `just_date` parameter in uniqueness checks, refines test coverage, and updates schema and documentation for new functionality.

### Domain wildcard replacement and parameter preprocessing

* Added a new `_preprocess_operation_params` method to `RuleProcessor` that performs shallow copying and recursive wildcard replacement for operation parameters, ensuring domain-specific variable names are used without mutating the original input objects. This method also handles supplemental domains by switching to `rdomain` when appropriate. (`cdisc_rules_engine/utilities/rule_processor.py`)
* Updated the rule processor to call `_preprocess_operation_params` before executing operations, ensuring all parameters are correctly formatted for the current domain. (`cdisc_rules_engine/utilities/rule_processor.py`)
* Refactored tests to cover the new preprocessing logic, including cases for wildcard replacement and supplemental domain handling. (`tests/unit/test_utilities/test_rule_processor.py`)

### Uniqueness check enhancements

* Enhanced the `is_unique_set` operator to support a new `just_date` parameter, which allows uniqueness checks to be performed on just the date portion of datetime values, ignoring time. (`cdisc_rules_engine/check_operators/dataframe_operators.py`)
* Updated the operator schema and documentation to include the new `just_date` parameter, clarifying its usage for users. (`resources/schema/Operator.json`, `resources/schema/Operator.md`) [[1]](diffhunk://#diff-6edd6c8cd18df7920523b202caf4fd52452dfff71e16f2328ab93e8066928c47L595-R596) [[2]](diffhunk://#diff-88f1326cb2643cd3f5193aaf1b6803a9376daca66f170d111a415d94ad6b3a49R825-R834)

### Minor fixes and test improvements

* Fixed variable naming in the `variable_is_null` operation to avoid unnecessary domain prefix replacement, simplifying the logic. (`cdisc_rules_engine/operations/variable_is_null.py`, `tests/unit/test_operations/test_variable_is_null.py`) [[1]](diffhunk://#diff-d47f4456097fbf1e39badb367472e0feebc84183f29c8db99e073caeb68ff0ddL17-R17) [[2]](diffhunk://#diff-763bfae2873d2111443045be82a6b260ce3abba7f408d8554bb7c90b1366826cL45-R45)
* Updated grouping logic in record count operations to use the correct dataset reference. (`cdisc_rules_engine/operations/record_count.py`)
* Refactored test setup to use consistent dataset metadata objects and improved test coverage for rule operations. (`tests/unit/test_utilities/test_rule_processor.py`) [[1]](diffhunk://#diff-85625e39b07f9388ff317713f2747de66c9433e69ff93aa0dd72a6fe9ec7e446R483-R497) [[2]](diffhunk://#diff-85625e39b07f9388ff317713f2747de66c9433e69ff93aa0dd72a6fe9ec7e446R578-R594) [[3]](diffhunk://#diff-85625e39b07f9388ff317713f2747de66c9433e69ff93aa0dd72a6fe9ec7e446R696-R712) [[4]](diffhunk://#diff-85625e39b07f9388ff317713f2747de66c9433e69ff93aa0dd72a6fe9ec7e446R757-R770)

These changes collectively improve the accuracy, maintainability, and usability of rule processing and dataset operations in the codebase.

### CODE REVIEW

File: cdisc_rules_engine/operations/record_count.py [hunk 1]
```diff
@@ -56,8 +56,8 @@ def _build_effective_grouping(self) -> tuple[list, dict]:
         )
         effective_grouping = []
         for col in grouping_cols:
-            if col in self.params.dataframe.columns:
-                sample_val = self.params.dataframe[col].iloc[0]
+            if col in self.evaluation_dataset.data.columns:
+                sample_val = self.evaluation_dataset[col].iloc[0]
```
Comment (RamilCDISC) [inline cdisc_rules_engine/operations/record_count.py:60]: I think it should `self.evaluation_dataset.data[col].iloc[0]` instead of `self.evaluation_dataset[col].iloc[0]`

File: cdisc_rules_engine/utilities/rule_processor.py [hunk 2]
```diff
@@ -599,6 +600,50 @@ def add_comparator_to_rule_conditions(
...
+        self, operation_params: OperationParams, domain_details: dict = None
+    ) -> OperationParams:
+        # uses shallow copy to not overwrite for subsequent
+        # operations and avoids costly deepcopy of dataframe
+        params_copy = copy.copy(operation_params)
+        current_domain = params_copy.domain
+        if domain_details.is_supp:
+            current_domain = domain_details.rdomain
+        for param_name in vars(params_copy):
+            if param_name == ("datasets" or "dataframe"):
```
Comment (RamilCDISC) [inline cdisc_rules_engine/utilities/rule_processor.py]: Here '==' will never match 'dataframe'. Instead we should use in.`if param_name in ("datasets" or "dataframe"):`

File: cdisc_rules_engine/check_operators/dataframe_operators.py [hunk 3]
```diff
@@ -1169,15 +1170,32 @@ def is_inconsistent_across_dataset(self, other_value):
...
         comparator = other_value.get("comparator")
+        just_date = other_value.get("just_date", False)
         values = [target, comparator]
         target_data = flatten_list(self.value, values)
         target_names = []
         for target_name in target_data:
             target_name = self.replace_prefix(target_name)
             if target_name in self.value.columns:
-                target_names.append(target_name)
+                if target_name.startswith("$"):
```
Comment (gerrycampion) [inline cdisc_rules_engine/check_operators/dataframe_operators.py]: what significance does "$" have? I know it is common practice to name operation id's this way, but i didn't think it was required to do this. should this branch happen regardless of the "$" prefix?

File: resources/schema/Operator.json [hunk 4]
```diff
@@ -592,7 +592,8 @@
     "type_insensitive": { "type": "boolean" },
     "round_values": { "type": "boolean" },
     "within": { "$ref": "CORE-base.json#/$defs/VariableName" },
-    "regex": { "type": "string" }
+    "regex": { "type": "string" },
+    "just_date": { "type": "boolean" }
```
Comment (gerrycampion) [inline resources/schema/Operator.json]: could we make this more generic by allowing a regex match instead of only supporting a date substring?

File: resources/schema/Operator.md
Comment (gerrycampion) [inline resources/schema/Operator.md:1]: remember to add to the rule editor params as well:
https://github.com/cdisc-org/conformance-rules-editor/blob/main/docs/check_parameter.md
We might want to think about pulling this doc over to the engine, but probably in another ticket.

File: tests/unit/test_operations/test_variable_is_null.py [hunk 5]
```diff
@@ -42,7 +42,7 @@ def test_variable_is_null(
     config = ConfigService()
     cache = CacheServiceFactory(config).get_cache_service()
     operation_params.dataframe = data
-    operation_params.target = "--VAR"
```
Comment (gerrycampion) [inline tests/unit/test_operations/test_variable_is_null.py:45]: Just confirming, "--VAR" doesn't work here anymore because the replacement is done in preprocessing which isn't a part of this unit test. I assume this is tested somewhere else instead?

File: cdisc_rules_engine/check_operators/dataframe_operators.py [hunk 6]
```diff
@@ -1169,15 +1170,32 @@ def is_inconsistent_across_dataset(self, other_value):
...
+                    list_val = self.value[target_name].iloc[0]
+                    if isinstance(list_val, (list, tuple)):
+                        target_names.extend(list_val)
+                    else:
+                        target_names.append(list_val)
+                else:
+                    target_names.append(target_name)
         target_names = list(set(target_names))
         df_group = self.value[target_names].copy()
+        if just_date:
```
Comment (gerrycampion) [inline cdisc_rules_engine/check_operators/dataframe_operators.py]: this should be added to the unit test

**Review by gerrycampion — CHANGES_REQUESTED** [review]

File: cdisc_rules_engine/check_operators/dataframe_operators.py (hunk 3 above)
Comment (SFJohnson24) [reply cdisc_rules_engine/check_operators/dataframe_operators.py]: this was/is deadcode--I changed this to use flattenlist() to extract the columns from the operation so it will never actually enter that logic but forgot to delete this conditional

File: cdisc_rules_engine/operations/record_count.py (hunk 1 above)
Comment (SFJohnson24) [reply cdisc_rules_engine/operations/record_count.py:60]: the iloc() method exists in both the Dask and PandasDataset classes
image
image
they produce the same thing

File: cdisc_rules_engine/utilities/rule_processor.py (hunk 2 above)
Comment (SFJohnson24) [reply cdisc_rules_engine/utilities/rule_processor.py]: updated

File: resources/schema/Operator.json (hunk 4 above)
Comment (SFJohnson24) [reply resources/schema/Operator.json]: updated to use regex string instead of the bool

File: resources/schema/Operator.md
Comment (SFJohnson24) [reply resources/schema/Operator.md:1]: done

File: tests/unit/test_operations/test_variable_is_null.py (hunk 5 above)
Comment (SFJohnson24) [reply tests/unit/test_operations/test_variable_is_null.py:45]: yes this doesnt work anymore due to the wildcard logic moving to the rule_processor.  I added unit tests to the rule_processor to show that it correctly changes wildcards to the current dataset for appropriate targeting

**Review by RamilCDISC — APPROVED** [review]
The PR updates the way '--' are handled and adds regex logic. The PR was validated by:

1. Reviewing the PR for any unwanted code or comment.
2. Reviewing the updated logic in accordance with AC.
3. Ensuring all unit and regression tests pass.
4. Ensuring all relevant testing is updated.
5. Validating the AC using positive datastes in dev editor.
6. Validating the AC using negative datasets in dev editor.
7. Validating the edge cases related to regex.
